cli/smartcontract/smart_contract.go

```diff
@@ -439,16 +447,30 @@ func invokeInternal(ctx *cli.Context, signAndPush bool) error {
 	paramsStart++
 
 	if len(args) > paramsStart {
-		for k, s := range args[paramsStart:] {
+		for k := paramsStart; k < len(args); {
```

Comment:
Maybe we can move this loop out into some function and add some tests for it? It gets more and more complex.

Reply:
And I think it can even be unified with `parseNestedArray`. It only differs in its `]` and `--` handling. Other than that they're identical.